### PR TITLE
reducing number of maxHits in CA

### PR DESCRIPTION
**Pull request by felicepantaleo**
The maximum number of hits per layer for PU50 rarely exceeds 4500 in BPix1. 
The limit was set to 20k and now reduced to 6k.

### CODE REVIEW

**Comment by felicepantaleo** [comment]
:construction: Validation running at vinavx2.cern.ch:/data/user/fpantale/patatrack-validation/run.Be59wxIF1T ...

**Comment by felicepantaleo** [comment]
:construction: Validation running at vinavx2.cern.ch:/data/user/fpantale/patatrack-validation/run.9on5a7ilkO ...

**Comment by fwyzard** [comment]
I agree that we should size the memory buffers to accommodate "most" of the events, rather than the possible theoretical maximum; however, we also need some way to handle the tails.

**Comment by fwyzard** [comment]
I've created #164 as a reminder.

**Comment by felicepantaleo** [comment]
the approach now has changed. 
We have only one global array of hits and `isOuterHitOfCell`
So this is now increased to the maximum sum of the averages per layer instead of the sum of the maximum sizes per layer

**Comment by felicepantaleo** [comment]
:construction: Validation running at vinavx2.cern.ch:/data/user/fpantale/patatrack-validation/run.Z0CbcjqUru ...

File: RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGeneratorGPU.cu [hunk 1]
```diff
@@ -124,9 +123,9 @@ void CAHitQuadrupletGeneratorGPU::allocateOnGPU()
   cudaCheck(cudaMemset(device_nCells_, 0, sizeof(uint32_t)));
 
   cudaCheck(cudaMalloc(&device_isOuterHitOfCell_,
-             maxNumberOfLayers_ * maxNumberOfHits_ * sizeof(GPU::VecArray<unsigned int, maxCellsPerHit_>)));
+             GPUConstants::maxNumberOfHits * sizeof(GPU::VecArray<unsigned int, maxCellsPerHit_>)));
```
Comment (fwyzard) [inline RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGeneratorGPU.cu]: I would add `Pixel` to `GPUConstants::maxNumberOfHits` somewere.

**Comment by fwyzard** [comment]
The plots do show some small differences in the fakes and duplicates.
Can we assume them to be within the usual noise ?

**Comment by felicepantaleo** [comment]
:construction: Validation running at vinavx2.cern.ch:/data/user/fpantale/patatrack-validation/run.P6ksQOlSkS ...

**Comment by fwyzard** [comment]
@felicepantaleo is this in good shape to be merged, or do you expect further changes ?

**Comment by felicepantaleo** [comment]
You can merge, I will make a more generic optimization PR once I'm back to
CERN

**Comment by fwyzard** [comment]
Here's the distribution of the number of hits for the first ~2800 events:
image

**Comment by fwyzard** [comment]
The distribution for data is reasonably a Gaussian at 18347 ± 3480, so `30000` is at ~3.3 σ from the average.
To be on the safe side, I'd bump to `40000` which is ~6 σ away...

**Comment by fwyzard** [comment]
With the latest changes, I am able to run on a larger sample of data.
Comparing the performance with and without this PR, I see no changes in the kernel themselves:

**before**
```
cuobjdump -arch=sm_70 ...
 Function kernel_connect(GPU::SimpleVector<Quadruplet>*, GPUCACell*, unsigned int const*, GPU::VecArray<unsigned int, 256>*, float, float, float, float, float, unsigned int, unsigned int):
  REG:48 STACK:8 SHARED:0 LOCAL:0 CONSTANT[0]:412 TEXTURE:0 SURFACE:0 SAMPLER:0
 Function kernel_checkOverflows(GPU::SimpleVector<Quadruplet>*, GPUCACell*, unsigned int const*, GPU::VecArray<unsigned int, 256>*, unsigned int):
  REG:24 STACK:8 SHARED:0 LOCAL:0 CONSTANT[0]:388 TEXTURE:0 SURFACE:0 SAMPLER:0

nvprof:
                       Time     Calls       Avg       Min       Max  Name
                   132.65ms      1200  110.54us  25.311us  385.15us  kernel_connect(...)
                    21.77ms      1200   18.14us   5.280us   34.24us  kernel_checkOverflows(...)
```

**after**
```
cuobjdump -arch=sm_70 ...
 Function kernel_connect(GPU::SimpleVector<Quadruplet>*, GPUCACell*, unsigned int const*, GPU::VecArray<unsigned int, 256>*, float, float, float, float, float, unsigned int, unsigned int):
  REG:48 STACK:8 SHARED:0 LOCAL:0 CONSTANT[0]:412 TEXTURE:0 SURFACE:0 SAMPLER:0
 Function kernel_checkOverflows(GPU::SimpleVector<Quadruplet>*, GPUCACell*, unsigned int const*, GPU::VecArray<unsigned int, 256>*, unsigned int, unsigned int):
  REG:24 STACK:8 SHARED:0 LOCAL:0 CONSTANT[0]:392 TEXTURE:0 SURFACE:0 SAMPLER:0

nvprof:
                       Time     Calls       Avg       Min       Max  Name
                   132.01ms      1200  110.01us  26.912us  392.12us  kernel_connect(...)
                    21.32ms      1200   17.76us   5.184us   36.77us  kernel_checkOverflows(...)
```

However, I do observe a significant boost to the overall throughput:

**before**
  - mean:   551 ± 20 ev/s
  - best:   573 ± 22 ev/s

**after**
  - mean:   639 ± 20 ev/s
  - best:   641 ± 26 ev/s

**Pull request merged**